File: 721-accounts-merge/721. Accounts Merge.cpp
class Solution {
    vector<int> id,sz;
    int find(int x){
        if(x==id[x])return x;
        return id[x]=find(id[x]);
    }
    void onion(int x, int y){
        int a = find(x);
        int b = find(y);
        if(a==b)return;
        if(sz[a]<sz[b])swap(a,b);
        id[b]=a;
        sz[a]+=sz[b];
    }
public:
    vector<vector<string>> accountsMerge(vector<vector<string>>& accounts) {
        int n = accounts.size();
        sz = vector<int>(n,1);
        id = vector<int>(n);
        map<string,int> a2i;
        for(int i = 0; i < n; i++){
            id[i]=i;
            for(int j = 1; j<accounts[i].size(); j++){
                auto it = a2i.find(accounts[i][j]);
                if(it==a2i.end()){
                    a2i[accounts[i][j]]=i;
                }else{
                    onion(it->second,i);
                }
            }
        }
        vector<vector<string>> ans;
        vector<int> i2a(n,-1);
        int pos;
        for(auto a2i_it:a2i){
            pos = find(a2i_it.second);
            if(i2a[pos]==-1){
                i2a[pos] = ans.size();
                ans.push_back({accounts[pos][0]});
                
            }
            ans[i2a[pos]].push_back(a2i_it.first);
        }

        return ans;
    }
};

File: 721-accounts-merge/721. Accounts Merge.py

class Solution:
    def accountsMerge(self, accounts: List[List[str]]) -> List[List[str]]:
        a2i = {}
        i2n = {}
        self.p = [_ for _ in range(len(accounts))]
        # size = [1 for _ in range(len(accounts))]
        def find(mail,i):
            nonlocal a2i
            return self.p[a2i.get(mail,i)]
        def u_find(i):
            if self.p[i]!=i:
                self.p[i]=u_find(self.p[i])
            return self.p[i]
        def union(a,b):
            if self.p[i]!=i:
                self.p[i2]=self.p[i]
            else:
                self.p[i]=self.p[i2]
        for i,account in enumerate(accounts):
            name = account[0]
            i2n[i] = name 
            for mail in account[1:]:
                i2 = find(mail,i)
                i2 = u_find(i2)
                if i2!=i:
                    union(i,i2)
                a2i[mail] = self.p[i]
        i2a = defaultdict(list)
        print(self.p)
        for a,i in a2i.items():
            i2a[u_find(i)].append(a)
        print(self.p)
        ans = []
        for i,a in i2a.items():
            ans.append([i2n[i],*sorted(a)])
        return ans
                
            
            